### PR TITLE
strings: Optimized ToTitle, tests and some cleanup

### PR DESCRIPTION
**Pull request by addvanced**
### Description:
I went through the strings package and did some cleanup in the tests.
Also, I optimised the "ToTitle" function - here's the benchstat:

#### ⏱ `sec/op`
| Exceptions | Before     | After      | Δ vs. base |
|------------|------------|------------|------------|
| nil        | 518.2n ± 2% | 394.8 ±1% | **-23.81%** |
| 0          | 526.4n ± 1% | 393.5n ± 1% | **-25.25%** |
| 1          | 573.1n ± 3% | 476.2n ± 0% | **-16.91%** |
| 2          | 556.0n ± 1% | 478.5n ± 1% | **-13.94%** |
| **geomean**| **543.0n**  | **433.8n** | **-20.11%** |

#### 📦 `B/op`
| Exceptions | Before     | After     | Δ vs. base |
|------------|------------|-----------|------------|
| nil        | 384.0 ± 0% | 296.0 ± 0%| **-22.92%** |
| 0          | 384.0 ± 0% | 296.0 ± 0%| **-22.92%** |
| 1          | 384.0 ± 0% | 296.0 ± 0%| **-22.92%** |
| 2          | 376.0 ± 0% | 296.0 ± 0%| **-21.28%** |
| **geomean**| **382.0**  | **296.0** | **-22.51%** |

#### 🗂 `allocs/op`
| Exceptions | Before     | After     | Δ vs. base |
|------------|------------|-----------|------------|
| nil        | 17.000 ± 0%| 3.000 ± 0%| **-82.35%** |
| 0          | 17.000 ± 0%| 3.000 ± 0%| **-82.35%** |
| 1          | 15.000 ± 0%| 3.000 ± 0%| **-80.00%** |
| 2          | 14.000 ± 0%| 3.000 ± 0%| **-78.57%** |
| **geomean**| **15.70**  | **3.000** | **-80.89%** |

### Checklist:  
- [x] **Tests Passing**: Verify by running `make test`.  
- [x] **Golint Passing**: Confirm by running `make lint`.  

#### If added a new utility function or updated existing function logic:
* [ ] Updated the utility package `EXAMPLES.md`
* [ ] Updated the utility package `README.md`

#### If added a new utility package:
* [ ] Updated the main `README.md` utility packages table

## Summary by CodeRabbit

* **Refactor**
  * Streamlined and optimized string utilities for better Unicode correctness, memory use, and performance without changing public APIs.
  * Removed an internal helper and consolidated logic to reduce intermediate allocations and simplify control flow.

* **Tests**
  * Added extensive new tests (including Unicode/emoji and edge cases), improved assertions, parallelized many subtests, and expanded benchmarks for clearer measurements.

* **Chores**
  * Minor cleanup and test/benchmark hygiene improvements.

### CODE REVIEW

File: strings/strings.go [hunk 1]
```diff
@@ -28,31 +28,29 @@ type TruncateOptions struct {
...
 			break
 		}
 
 		if options.ReturnIndexes {
-			result = append(result, input[index+startIndex:])
+			result = append(result, input[start+idx:])
 		} else {
-			result = append(result, input[startIndex+index:startIndex+index+len(substring)])
+			result = append(result, input[start+idx:start+idx+len(substring)])
 		}
```
[the remark this answers is not in the document]
Comment (addvanced) [reply strings/strings.go:53]: @kashifkhan0771 So, as coderabbitai mentions, return the index as a string?

File: strings/strings.go [hunk 2]
```diff
@@ -66,25 +64,49 @@ func Title(input string) string {
...
 
-	// Create a maps of exceptions for faster lookup
-	exceptionMap := make(map[string]bool)
-	for _, exception := range exceptions {
-		exceptionMap[exception] = true
+	// O(1) lookup map for exceptions
+	exc := make(map[string]struct{}, len(exceptions))
+	for _, e := range exceptions {
+		exc[e] = struct{}{}
 	}
```
[the remark this answers is not in the document]
Comment (addvanced) [reply strings/strings.go:77]: @kashifkhan0771 I just committed some changes. Let me know what you think. If you accept that solution, let me know if I should delete the helper.go file with the `firstLetterToUpper` function inside, since it is no longer used.

[the remark this answers is not in the document]
Comment (addvanced) [reply strings/strings.go:77]: and also, I will update the performance benchmarks in the description. This change had an impact on the performance, but it's still better than the original code.

File: strings/strings.go [hunk 3]
```diff
@@ -226,19 +251,15 @@ func CaesarDecrypt(input string, shift int) string {
...
+	const pattern = `^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$`
 
 	isValid, err := regexp.MatchString(pattern, email)
-	if err != nil {
-		return false
-	}
 
-	return isValid
+	return err == nil && isValid
 }
```
[the remark this answers is not in the document]
Comment (addvanced) [reply strings/strings.go]: Implemented this. It obviously makes sense to be unicode compatible, just like we want it to be in the ToTitle() function.

File: strings/strings.go [hunk 4]
```diff
@@ -28,31 +30,32 @@ type TruncateOptions struct {
...
+			result = append(result, strconv.Itoa(start+idx))
 		} else {
-			result = append(result, input[startIndex+index:startIndex+index+len(substring)])
+			result = append(result, input[start+idx:start+idx+len(substring)])
 		}
-
-		startIndex += index + 1
 	}
 
 	return result
```
[the remark this answers is not in the document]
Comment (addvanced) [reply strings/strings.go]: Implemented this. It obviously makes sense to be unicode compatible, just like we want it to be in the ToTitle() function.

File: strings/strings.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (addvanced) [reply strings/strings.go]: Done.

**Review by kashifkhan0771 — APPROVED** [review]
Thank you @addvanced ❤️
